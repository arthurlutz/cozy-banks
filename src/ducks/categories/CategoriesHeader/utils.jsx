export const makeBreadcrumbs = (router, categoryName, subcategoryName, t) => {
  const breadcrumbs = [
    {
      name: t('Categories.title.general'),
      onClick: () => router.push('/analysis/categories')
    }
  ]
  if (categoryName) {
    breadcrumbs.push({
      name: t(`Data.categories.${categoryName}`),
      onClick: () => router.push(`/analysis/categories/${categoryName}`)
    })
  }
  if (subcategoryName) {
    breadcrumbs.push({
      name: t(`Data.subcategories.${subcategoryName}`)
    })
  }
  return breadcrumbs
}
