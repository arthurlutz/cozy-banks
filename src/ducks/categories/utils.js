const onSubcategory = params => Boolean(params.subcategoryName)

export { onSubcategory }
