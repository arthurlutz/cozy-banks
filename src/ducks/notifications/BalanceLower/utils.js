export const hasOneAccount = templateData => {
  const { accounts } = templateData
  const hasOneAccount = accounts.length === 1
  return hasOneAccount
}
