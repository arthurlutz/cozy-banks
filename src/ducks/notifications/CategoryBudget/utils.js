import isEqual from 'lodash/isEqual'
import logger from 'cozy-logger'

import { collectAlertInfo } from 'ducks/budgetAlerts'
import CategoryBudgetNotificationView from 'ducks/notifications/CategoryBudget'

const lang = process.env.COZY_LOCALE || 'en'
const dictRequire = lang => require(`locales/${lang}`)

const log = logger.namespace('category-alerts')

/**
 * Collects notification data for all alerts
 *
 * Return nulls if nothing is to be sent
 */
export const buildNotificationData = async (client, alerts, options = {}) => {
  if (alerts.length === 0) {
    log('info', 'No category budget alerts, bailing out.')
  }

  const data = []
  for (let alert of alerts) {
    const info = { alert }
    try {
      const collectedInfo = await collectAlertInfo(client, alert, options)
      if (collectedInfo) {
        Object.assign(info, collectedInfo)
      }
    } catch (e) {
      log(
        'error',
        `Error while checking budget alert ${alert.id} (message: ${e.message})`
      )
    }
    data.push(info)
  }

  const updatedAlerts = data.map(x => x.alert)
  if (isEqual(alerts, updatedAlerts) && !options.force) {
    log('info', 'No change to alerts, no need to send')
    return null
  }

  return data
}

export const buildNotificationView = (client, options) => {
  const notifView = new CategoryBudgetNotificationView({
    client,
    lang,
    data: {},
    locales: {
      [lang]: dictRequire(lang)
    },
    ...options
  })
  return notifView
}

export const hasBudgetAlerts = templateData => {
  const { budgetAlerts } = templateData
  const hasMultipleAlerts = budgetAlerts.length > 1
  return hasMultipleAlerts
}
