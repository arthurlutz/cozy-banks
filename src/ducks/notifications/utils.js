import groupBy from 'lodash/groupBy'
import sortBy from 'lodash/sortBy'
import toPairs from 'lodash/toPairs'
import flow from 'lodash/flow'
import unique from 'lodash/uniq'
import get from 'lodash/get'
import { getDate } from 'ducks/transactions/helpers'

export const prepareTransactions = function(transactions) {
  const byAccounts = groupBy(transactions, tr => tr.account)

  const groupAndSortByDate = flow(
    transactions => groupBy(transactions, getDate),
    toPairs,
    dt => sortBy(dt, ([date]) => date).reverse()
  )
  Object.keys(byAccounts).forEach(account => {
    byAccounts[account] = groupAndSortByDate(byAccounts[account])
  })

  return byAccounts
}

const billIdFromReimbursement = reimbursement => {
  return reimbursement.billId && reimbursement.billId.split(':')[1]
}

export const treatedByFormat = function(reimbursements, billsById) {
  if (!billsById) {
    throw new Error('No billsById passed')
  }
  const vendors = unique(
    (reimbursements || [])
      .map(reimbursement => {
        const billId = billIdFromReimbursement(reimbursement)
        return get(billsById, billId + '.vendor')
      })
      .filter(x => x && x !== '')
  )

  if (!vendors.length) {
    throw new Error('No vendor found')
  }
  return vendors.join(', ')
}

export const getCurrentDate = () => {
  return new Date()
}

export const formatAmount = amount =>
  amount % 1 !== 0 ? amount.toFixed(2) : amount
