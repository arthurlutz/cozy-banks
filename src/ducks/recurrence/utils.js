import uniqBy from 'lodash/uniqBy'
import startCase from 'lodash/startCase'
import groupBy from 'lodash/groupBy'
import addDays from 'date-fns/add_days'
import parse from 'date-fns/parse'
import differenceInDays from 'date-fns/difference_in_days'

import {
  getCategoryId,
  getLabel as getTransactionLabel
} from 'ducks/transactions/helpers'
import { brandSplit, mergeBundles } from './rules'

const RECURRENCE_DOCTYPE = 'io.cozy.bank.recurrence'

export const prettyLabel = label => {
  return label ? startCase(label.toLowerCase()) : ''
}

export const getLabel = bundle => {
  if (bundle.manualLabel) {
    return bundle.manualLabel
  } else {
    return prettyLabel(bundle.automaticLabel)
  }
}

export const getCategories = bundle => {
  return bundle.categoryIds
}

export const getFrequencyWord = freq => {
  if (freq > 5 && freq <= 8) {
    return 'weekly'
  } else if (freq > 26 && freq <= 40) {
    return 'monthly'
  } else if (freq > 300 && freq <= 400) {
    return 'yearly'
  }
}

export const getFrequency = recurrence => {
  return recurrence.stats.deltas.median
}

export const getFrequencyText = (t, recurrence) => {
  const frequency = Math.floor(getFrequency(recurrence))
  const freqWord = getFrequencyWord(frequency)
  const text = freqWord
    ? t(`Recurrence.freq-info-word.${freqWord}`)
    : t(`Recurrence.freq-info`, { frequency })
  return text
}

export const getAmount = bundle => {
  return bundle.amounts[0]
}

export const getCurrency = () => {
  return 'â‚¬'
}

/**
 * Make a recurrence from a transaction. The recurrence will inherit
 * the amount / account / category of the transaction
 *
 * @param  {Transaction} transaction (not hydrated)
 * @return {Recurrence}
 */
export const makeRecurrenceFromTransaction = transaction => {
  const accountId = transaction.account
  return {
    _type: RECURRENCE_DOCTYPE,
    automaticLabel: getTransactionLabel(transaction),
    stats: {
      deltas: {
        median: 30
      }
    },
    latestDate: transaction.date,
    accounts: [accountId],
    amounts: [transaction.amount],
    categoryIds: [getCategoryId(transaction)]
  }
}

/** Gives the next date for a recurrence, based on the median stat and the latest date */
export const nextDate = recurrence => {
  try {
    const {
      latestDate: rawLatestDate,
      stats: {
        deltas: { median }
      }
    } = recurrence
    const latestDate = parse(rawLatestDate)
    const now = new Date(Date.now())
    const deltaDays = differenceInDays(now, latestDate)
    const n = deltaDays / median
    const r = Math.ceil(n)
    const candidate = addDays(latestDate, r * median)
    if (differenceInDays(candidate, now) <= 0) {
      return new Date(now)
    }
    return candidate
  } catch (e) {
    // eslint-disable-next-line
    console.error('Error while computing next date', e)
    return null
  }
}

export const isDeprecatedBundle = recurrence => {
  const latestDate = parse(recurrence.latestDate)
  const now = Date.now()
  return differenceInDays(now, latestDate)
}

export const addTransactionToBundles = (bundles, transactions) => {
  const updatedBundles = [...bundles].map(b => {
    const bundle = { ...b }

    // Matching on Amount, CategoryId and account
    const transactionFounds = transactions.filter(transaction => {
      return (
        bundle.categoryIds.some(
          catId => getCategoryId(transaction) === catId
        ) &&
        bundle.amounts.some(amount => amount === transaction.amount) &&
        bundle.accounts?.some(account => account === transaction.account)
      )
    })

    if (transactionFounds?.length > 0) {
      bundle.ops = uniqBy([...bundle.ops, ...transactionFounds], o => o._id)
    }

    return bundle
  })

  return updatedBundles
}

const findBrandBundles = transactions => {
  const brandBundles = transactions
    .map(t => ({
      ops: [t],
      categoryIds: [getCategoryId(t)],
      amounts: [t.amount]
    }))
    .map(brandSplit())
    .map(b => b[0])
    .filter(b => Boolean(b.brand))
    .filter(b => Boolean(b.categoryIds[0]))

  return brandBundles
}

export const mergeBrandBundles = transactions => {
  const brandBundles = findBrandBundles(transactions)

  const groups = groupBy(brandBundles, b => {
    return `${b.categoryIds[0]}/${b.amounts[0]}/${b.brand}`
  })
  const mergedBundles = Object.values(groups).map(mergeBundles)
  return mergedBundles
}
