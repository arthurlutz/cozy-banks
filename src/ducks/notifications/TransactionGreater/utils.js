import { getCurrencySymbol } from 'utils/currencySymbol'
import capitalize from 'lodash/capitalize'

export const MAX_CHAR_BY_LINE = 50
const SEPARATOR = ' : '

const capitalizeEachWords = str =>
  str
    .split(' ')
    .map(capitalize)
    .join(' ')

export const formatTransaction = transaction => {
  const { amount, currency } = transaction
  const amountFormat = `${amount}${getCurrencySymbol(currency)}`.replace(
    '.',
    ','
  )
  const REMAINS_LENGTH =
    MAX_CHAR_BY_LINE - SEPARATOR.length - amountFormat.length
  const labelCapitalize = capitalizeEachWords(transaction.label)
  const transactionName = labelCapitalize.substr(0, REMAINS_LENGTH)
  return `${transactionName}${SEPARATOR}${amountFormat}`
}
