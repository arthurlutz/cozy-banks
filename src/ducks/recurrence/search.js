import groupBy from 'lodash/groupBy'
import unique from 'lodash/uniq'
import uniq from 'lodash/uniq'
import compose from 'lodash/flowRight'
import maxBy from 'lodash/maxBy'
import minBy from 'lodash/maxBy'
import flatMap from 'lodash/flatMap'

import defaultRulesConfig from './config.json'
import {
  addStats,
  getRulesFromConfig,
  groupBundles,
  overEvery
} from 'ducks/recurrence/rules'
import getCategoryId from 'ducks/transactions/getCategoryId'

import { getLabel } from 'ducks/transactions/helpers'
import { addTransactionToBundles } from 'ducks/recurrence/utils'

const ONE_DAY = 86400 * 1000

const assert = (pred, msg) => {
  if (!pred) {
    throw new Error(msg)
  }
}

/**
 * How rules work:
 *
 * There are different stages that find and process bundles
 *
 * - Some stages filter bundles
 * - Some change the bundles (map)
 * - Some regroup bundles
 *
 * @param  {array} operations
 * @param  {array} rules
 * @return {array} recurrence groups
 */
export const findRecurrences = (operations, rules) => {
  const groups = groupBy(operations, x => getCategoryId(x))

  let bundles = flatMap(Object.entries(groups), ([categoryId, ops]) => {
    const perAmount = groupBy(ops, op => op.amount)
    return Object.entries(perAmount).map(([amount, ops]) => ({
      categoryIds: [categoryId],
      amounts: [parseInt(amount, 10)],
      ops,
      automaticLabel: getLabel(ops[0])
    }))
  })

  const groupedByStage = groupBy(rules, rule => rule.stage)

  const stageKeys = Object.keys(groupedByStage).sort()
  for (let stageKey of stageKeys) {
    const ruleInfos = groupedByStage[stageKey]
    assert(
      unique(ruleInfos.map(r => r.type)).length === 1,
      'Cannot have multiple types per stage'
    )
    const type = ruleInfos[0].type
    const rules = ruleInfos.map(ruleInfo => ruleInfo.rule)
    if (type === 'filter') {
      bundles = bundles.filter(overEvery(rules))
    } else if (type === 'map') {
      bundles = bundles.map(compose(rules))
    } else if (type === 'group') {
      bundles = groupBundles(bundles, rules)
    } else if (type === 'split') {
      if (rules.length > 1) {
        throw new Error('Cannot have multiple split rules in one stage')
      }
      bundles = flatMap(bundles, rules[0])
    }
  }

  return bundles
}

export const updateRecurrences = (bundles, newTransactions, rules) => {
  if (!newTransactions.length) {
    return bundles
  }
  const maxDate = new Date(maxBy(newTransactions, 'date').date)
  const minDate = new Date(minBy(newTransactions, 'date').date)
  const dateSpan = (maxDate - minDate) / ONE_DAY

  let newBundles = []
  let updatedBundles
  if (dateSpan > 90 && newTransactions.length > 100) {
    newBundles = findRecurrences(newTransactions, rules)
  } else {
    updatedBundles = addTransactionToBundles(bundles, newTransactions)
  }

  const allBundles = [...updatedBundles, ...newBundles].map(addStats)
  return allBundles
}

export const findAndUpdateRecurrences = (recurrences, operations) => {
  const rules = getRulesFromConfig(defaultRulesConfig)

  let updatedRecurrences
  if (recurrences.length === 0) {
    updatedRecurrences = findRecurrences(operations, rules)
  } else {
    updatedRecurrences = updateRecurrences(recurrences, operations, rules)
  }
  return updatedRecurrences
}

export const updateAmountsCategoriesRecurrences = bundles => {
  const newBundles = [...bundles].map(b => {
    return {
      ...b,
      amounts: uniq(b.ops.map(o => o.amount)),
      categoryIds: uniq(b.ops.map(getCategoryId))
    }
  })
  return newBundles
}
