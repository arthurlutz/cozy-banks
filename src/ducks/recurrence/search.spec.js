import sortBy from 'lodash/sortBy'
import minBy from 'lodash/minBy'
import maxBy from 'lodash/maxBy'
import uniqBy from 'lodash/uniqBy'
import isArray from 'lodash/isArray'
import keyBy from 'lodash/keyBy'

import { TRANSACTION_DOCTYPE } from 'doctypes'
import { td, tr } from 'txt-table-utils'

import { findAndUpdateRecurrences } from './search'
import { getLabel, getAmount, getFrequencyText } from './utils'
import fixtures from './fixtures/fixtures.json'
import fixtures2 from './fixtures/fixtures2.json'
import fixtures3 from './fixtures/fixtures3.json'
import fixtures4 from './fixtures/fixtures4.json'
import fixtures5 from './fixtures/fixtures5.json'
import fixtures6 from './fixtures/fixtures6.json'
import { getT, enLocaleOption } from 'utils/lang'

const formatBundleExtent = bundle => {
  const oldestOp = minBy(bundle.ops, x => x.date)
  const latestOp = maxBy(bundle.ops, x => x.date)
  return `from ${oldestOp.date.slice(0, 10)} to ${latestOp.date.slice(0, 10)}`
}

const assert = (cond, error) => {
  if (!cond) {
    throw new Error(error)
  }
}

const assertUniqueOperations = recurrence => {
  const uniquedOps = uniqBy(recurrence.ops, op => op._id)
  assert(
    recurrence.ops.length === uniquedOps.length,
    `Duplicate ops in recurrence ${getLabel(recurrence)}`
  )
}

const assertValidRecurrence = bundle => {
  assertUniqueOperations(bundle)
  assert(isArray(bundle.amounts), 'Bundle should have amounts array')
  assert(isArray(bundle.categoryIds), 'Bundle should have categoryIds array')
}

const formatRecurrence = bundle =>
  tr(
    td(getLabel(bundle), 50),
    td(getAmount(bundle), 10, 'right'),
    td(`${bundle.ops.length} operations`, 17, 'right'),
    td(formatBundleExtent(bundle), 30, 'right'),
    td(bundle.stats.deltas.median, 7, 'right')
  )

describe('recurrence bundles', () => {
  it('should find new bundles', () => {
    const transactions = fixtures[TRANSACTION_DOCTYPE]
    const recurrences = []
    const updatedRecurrences = findAndUpdateRecurrences(
      recurrences,
      transactions
    )

    updatedRecurrences.forEach(assertValidRecurrence)
    // eslint-disable
    expect(
      sortBy(updatedRecurrences.map(formatRecurrence)).join('\n')
    ).toMatchSnapshot()
  })

  it('should find new bundles (split brand necessary)', () => {
    const transactions = fixtures2[TRANSACTION_DOCTYPE]
    const recurrences = []
    const updatedRecurrences = findAndUpdateRecurrences(
      recurrences,
      transactions
    )

    updatedRecurrences.forEach(assertValidRecurrence)
    // eslint-disable
    expect(
      sortBy(updatedRecurrences.map(formatRecurrence)).join('\n')
    ).toMatchSnapshot()
  })

  it('should find new bundles', () => {
    const transactions = fixtures3[TRANSACTION_DOCTYPE]
    const recurrences = []
    const updatedRecurrences = findAndUpdateRecurrences(
      recurrences,
      transactions
    )

    updatedRecurrences.forEach(assertValidRecurrence)
    // eslint-disable
    expect(
      sortBy(updatedRecurrences.map(formatRecurrence)).join('\n')
    ).toMatchSnapshot()
  })

  it('should update existing bundles', () => {
    const transactions = fixtures[TRANSACTION_DOCTYPE].filter(
      tr => tr.date < '2019-05'
    )
    const newTransactions = fixtures[TRANSACTION_DOCTYPE].filter(
      tr => tr.date > '2019-05'
    )
    const recurrences = findAndUpdateRecurrences([], transactions)

    expect(
      sortBy(recurrences.map(formatRecurrence)).join('\n')
    ).toMatchSnapshot()

    const updatedRecurrences = findAndUpdateRecurrences(
      recurrences,
      newTransactions
    )

    updatedRecurrences.forEach(assertValidRecurrence)

    expect(
      sortBy(updatedRecurrences.map(formatRecurrence)).join('\n')
    ).toMatchSnapshot()
  })
})

describe('recurrence scenario with all operations which have a subcategory', () => {
  const transactionsByKey = keyBy(fixtures4[TRANSACTION_DOCTYPE], '_id')
  const transactions = [
    transactionsByKey['february'],
    transactionsByKey['march'],
    transactionsByKey['april']
  ]
  const mayTransaction = transactionsByKey['may']
  const juneTransaction = transactionsByKey['june']

  const recurrence = {
    categoryIds: ['200110'],
    amounts: [2000, 2150],
    accounts: ['1d22740c6c510e5368d1b6b670deee05'],
    ops: [...transactions, mayTransaction],
    automaticLabel: 'Mon Salaire'
  }

  it('should match 01/06 salary with 2000 and salary in subcategory', () => {
    juneTransaction.amount = 2000

    const bundleWithJune = findAndUpdateRecurrences(
      [recurrence],
      [juneTransaction]
    )

    const bundle = bundleWithJune[0]
    const [op1, op2, op3, op4, op5] = bundle.ops
    expect(bundleWithJune.length).toBe(1)
    expect(bundle.categoryIds[0]).toBe('200110')
    expect(bundle.amounts.length).toBe(2)
    expect(bundle.amounts[0]).toBe(2000)
    expect(bundle.amounts[1]).toBe(2150)
    expect(bundle.ops.length).toBe(5)
    expect(op1._id).toBe('february')
    expect(op1.amount).toBe(2000)
    expect(op2._id).toBe('march')
    expect(op2.amount).toBe(2000)
    expect(op3._id).toBe('april')
    expect(op3.amount).toBe(2000)
    expect(op4._id).toBe('may')
    expect(op4.amount).toBe(2150)
    expect(op5._id).toBe('june')
    expect(op5.amount).toBe(2000)
  })

  it('should match 01/06 salary with 2150 and salary in subcategory', () => {
    juneTransaction.amount = 2150

    const bundleWithJune = findAndUpdateRecurrences(
      [recurrence],
      [juneTransaction]
    )

    const bundle = bundleWithJune[0]
    const [op1, op2, op3, op4, op5] = bundle.ops
    expect(bundleWithJune.length).toBe(1)
    expect(bundle.categoryIds[0]).toBe('200110')
    expect(bundle.amounts.length).toBe(2)
    expect(bundle.amounts[0]).toBe(2000)
    expect(bundle.amounts[1]).toBe(2150)
    expect(bundle.ops.length).toBe(5)
    expect(op1._id).toBe('february')
    expect(op1.amount).toBe(2000)
    expect(op2._id).toBe('march')
    expect(op2.amount).toBe(2000)
    expect(op3._id).toBe('april')
    expect(op3.amount).toBe(2000)
    expect(op4._id).toBe('may')
    expect(op4.amount).toBe(2150)
    expect(op5._id).toBe('june')
    expect(op5.amount).toBe(2150)
  })

  it('should not match 01/06 salary with 2100 and salary in subcategory', () => {
    juneTransaction.amount = 2100

    const bundleWithJune = findAndUpdateRecurrences(
      [recurrence],
      [juneTransaction]
    )

    expect(bundleWithJune.length).toBe(1)

    const bundle = bundleWithJune[0]
    const [op1, op2, op3, op4] = bundle.ops
    // Bundle #1
    expect(bundle.categoryIds[0]).toBe('200110')
    expect(bundle.amounts.length).toBe(2)
    expect(bundle.amounts[0]).toBe(2000)
    expect(bundle.amounts[1]).toBe(2150)
    expect(bundle.ops.length).toBe(4)
    expect(op1._id).toBe('february')
    expect(op1.amount).toBe(2000)
    expect(op2._id).toBe('march')
    expect(op3.amount).toBe(2000)
    expect(op3._id).toBe('april')
    expect(op3.amount).toBe(2000)
    expect(op4._id).toBe('may')
    expect(op4.amount).toBe(2150)
  })

  it('should have todoooo match 01/06 salary with 2100 and salary in subcategory with brands', () => {
    juneTransaction.amount = 2100

    const juneSpot = transactionsByKey['june-spotify']
    const julySpot = transactionsByKey['july-spotify']
    const recurrenceSpot = {
      brand: 'Spotify',
      categoryIds: ['400100'],
      amounts: [10],
      accounts: ['1d22740c6c510e5368d1b6b670deee05'],
      ops: [juneSpot],
      automaticLabel: 'Spotify Abonnement'
    }

    const bundles = sortBy(
      findAndUpdateRecurrences([recurrence, recurrenceSpot], [julySpot]),
      b => b.brand
    )

    expect(bundles.length).toBe(2)
    const bundle = bundles[0]

    expect(bundle.brand).toBe('Spotify')
    expect(bundle.categoryIds[0]).toBe('400100')
    expect(bundle.amounts.length).toBe(1)
    expect(bundle.amounts[0]).toBe(10)
    expect(bundle.ops.length).toBe(2)

    const bundle2 = bundles[1]
    const [op1, op2, op3, op4] = bundle2.ops
    expect(bundle2.brand).toBe(undefined)
    expect(bundle2.categoryIds[0]).toBe('200110')
    expect(bundle2.amounts.length).toBe(2)
    expect(bundle2.amounts[0]).toBe(2000)
    expect(bundle2.amounts[1]).toBe(2150)
    expect(bundle2.ops.length).toBe(4)
    expect(op1._id).toBe('february')
    expect(op1.amount).toBe(2000)
    expect(op2._id).toBe('march')
    expect(op2.amount).toBe(2000)
    expect(op3._id).toBe('april')
    expect(op3.amount).toBe(2000)
    expect(op4._id).toBe('may')
    expect(op4.amount).toBe(2150)
  })
})

describe('recurrence scenario with 01 feb, march and april are to categorize (0)', () => {
  const transactionsByKey = keyBy(fixtures5[TRANSACTION_DOCTYPE], '_id')
  const transactions = [
    transactionsByKey['february'],
    transactionsByKey['march'],
    transactionsByKey['april']
  ]

  const mayTransaction = transactionsByKey['may']
  const juneTransaction = transactionsByKey['june']

  const recurrence = {
    brand: '',
    categoryIds: ['0', '200110'],
    amounts: [2000, 2150],
    accounts: ['1d22740c6c510e5368d1b6b670deee05'],
    ops: [...transactions, mayTransaction],
    automaticLabel: 'Mon Salaire'
  }

  it('01/06 salary with 2000 and subcategory: to categorize', () => {
    juneTransaction.amount = 2000
    juneTransaction.automaticCategoryId = '0'

    const bundleWithJune = findAndUpdateRecurrences(
      [recurrence],
      [juneTransaction]
    )

    const bundle = bundleWithJune[0]
    const [op1, op2, op3, op4, op5] = bundle.ops
    expect(bundleWithJune.length).toBe(1)
    expect(bundle.categoryIds[0]).toBe('0')
    expect(bundle.categoryIds[1]).toBe('200110')
    expect(bundle.amounts.length).toBe(2)
    expect(bundle.amounts[0]).toBe(2000)
    expect(bundle.amounts[1]).toBe(2150)
    expect(bundle.ops.length).toBe(5)
    expect(op1._id).toBe('february')
    expect(op1.amount).toBe(2000)
    expect(op2._id).toBe('march')
    expect(op2.amount).toBe(2000)
    expect(op3._id).toBe('april')
    expect(op3.amount).toBe(2000)
    expect(op4._id).toBe('may')
    expect(op4.amount).toBe(2150)
    expect(op5._id).toBe('june')
    expect(op5.amount).toBe(2000)
  })

  it('01/06 salary with 2000 and salary in subcategory', () => {
    juneTransaction.amount = 2000
    juneTransaction.automaticCategoryId = '200110'

    const bundleWithJune = findAndUpdateRecurrences(
      [recurrence],
      [juneTransaction]
    )

    const bundle = bundleWithJune[0]
    const [op1, op2, op3, op4, op5] = bundle.ops
    expect(bundleWithJune.length).toBe(1)
    expect(bundle.categoryIds[0]).toBe('0')
    expect(bundle.categoryIds[1]).toBe('200110')
    expect(bundle.amounts.length).toBe(2)
    expect(bundle.amounts[0]).toBe(2000)
    expect(bundle.amounts[1]).toBe(2150)
    expect(bundle.ops.length).toBe(5)
    expect(op1._id).toBe('february')
    expect(op1.amount).toBe(2000)
    expect(op2._id).toBe('march')
    expect(op2.amount).toBe(2000)
    expect(op3._id).toBe('april')
    expect(op3.amount).toBe(2000)
    expect(op4._id).toBe('may')
    expect(op4.amount).toBe(2150)
    expect(op5._id).toBe('june')
    expect(op5.amount).toBe(2000)
  })

  it('reattach next new operations from a new bundle which has 1 operation with "very month" in frequency', () => {
    const transactionsByKey = keyBy(fixtures6[TRANSACTION_DOCTYPE], '_id')
    const febTransactionEDF = transactionsByKey['february']
    const mayTransactionEDF = transactionsByKey['may']

    const recurrence = {
      categoryIds: ['401080'],
      amounts: [30],
      accounts: ['1d22740c6c510e5368d1b6b670deee05'],
      ops: [febTransactionEDF],
      automaticLabel: 'EDF'
    }

    const bundleWithEDF = findAndUpdateRecurrences(
      [recurrence],
      [mayTransactionEDF]
    )

    const bundle = bundleWithEDF[0]
    expect(bundle.ops.length).toBe(2)

    const t = getT(enLocaleOption)
    expect(getFrequencyText(t, bundle)).toBe('every month')
    const [op1, op2] = bundle.ops
    expect(op1._id).toBe('february')
    expect(op1.amount).toBe(30)
    expect(op2._id).toBe('may')
    expect(op2.amount).toBe(30)
  })
})
