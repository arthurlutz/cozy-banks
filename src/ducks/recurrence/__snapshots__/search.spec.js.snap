// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recurrence bundles should find new bundles (split brand necessary) 1`] = `
"75 Netflix Com                                     |     -11.99 |     10 operations |  from 2020-09-01 to 2021-06-01 |      30
Spotify P 127537 A 66                              |     -14.99 |      7 operations |  from 2020-10-01 to 2021-04-01 |      30"
`;

exports[`recurrence bundles should find new bundles 1`] = `
"Cozy Cloud                                         |    3963.13 |      3 operations |  from 2018-06-01 to 2018-09-28 |    59.5
Echeance Pret 03103 61893938                       |   -1281.82 |     11 operations |  from 2019-03-05 to 2020-01-06 |      31
Edf Clients Particuliers                           |      -36.6 |     11 operations |  from 2018-08-24 to 2019-06-24 |      31
Free Mobile                                        |     -17.99 |     15 operations |  from 2018-05-22 to 2020-01-20 |      30
Free Telecom                                       |     -31.98 |     20 operations |  from 2018-05-07 to 2020-02-06 |      30
Navigo Annuel Gie Comutitres                       |      -75.2 |     20 operations |  from 2018-06-04 to 2020-03-03 |      31
Paypal Europe S A R L Et Cie S C A Ech             |     -13.99 |      4 operations |  from 2018-05-29 to 2018-10-30 |      32
Vr Permanent Bourso                                |        -60 |     22 operations |  from 2018-06-04 to 2020-03-02 |      31
Vr Permanent Loyer Douillet                        |    -913.01 |     11 operations |  from 2018-06-04 to 2019-04-03 |      31"
`;

exports[`recurrence bundles should find new bundles 2`] = `"Sepa Mlle Durand Genevieve                         |        -10 |     12 operations |  from 2020-11-30 to 2021-05-31 |       0"`;

exports[`recurrence bundles should update existing bundles 1`] = `
"Cozy Cloud                                         |    3963.13 |      3 operations |  from 2018-06-01 to 2018-09-28 |    59.5
Edf Clients Particuliers                           |      -36.6 |      9 operations |  from 2018-08-24 to 2019-04-24 |      31
Free Telecom                                       |     -31.98 |     12 operations |  from 2018-05-07 to 2019-04-05 |      30
Navigo Annuel Gie Comutitres                       |      -75.2 |     10 operations |  from 2018-06-04 to 2019-04-03 |      31
Paypal Europe S A R L Et Cie S C A Ech             |     -13.99 |      4 operations |  from 2018-05-29 to 2018-10-30 |      32
Vr Permanent Bourso                                |        -60 |     11 operations |  from 2018-06-04 to 2019-04-02 |    30.5
Vr Permanent Loyer Douillet                        |    -913.01 |     11 operations |  from 2018-06-04 to 2019-04-03 |      31"
`;

exports[`recurrence bundles should update existing bundles 2`] = `
"Cozy Cloud                                         |    3963.13 |      3 operations |  from 2018-06-01 to 2018-09-28 |    59.5
Edf Clients Particuliers                           |      -36.6 |      9 operations |  from 2018-08-24 to 2019-04-24 |      31
Free Telecom                                       |     -31.98 |     12 operations |  from 2018-05-07 to 2019-04-05 |      30
Navigo Annuel Gie Comutitres                       |      -75.2 |     10 operations |  from 2018-06-04 to 2019-04-03 |      31
Paypal Europe S A R L Et Cie S C A Ech             |     -13.99 |      4 operations |  from 2018-05-29 to 2018-10-30 |      32
Vr Permanent Bourso                                |        -60 |     11 operations |  from 2018-06-04 to 2019-04-02 |    30.5
Vr Permanent Loyer Douillet                        |    -913.01 |     11 operations |  from 2018-06-04 to 2019-04-03 |    30.5"
`;
